packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -0,0 +1,387 @@
+///         1. `vm.getDeployedBytecode` can only be used with `vm.etch` when there are no side
+///         effects in the constructor and no immutables in the bytecode.
+///         2. A contract must be deployed using the `new` syntax if there are immutables in the code.
+///         Any other side effects from the init code besides setting the immutables must be cleaned up afterwards.
+///         3. A contract is deployed using the `new` syntax because it's not proxied, but still needs to be set
+///         at a specific address. Because just deploying a new instance doesn't give us the contract at our desired
+/// address,
+///         we must use `vm.etch` to set the deployed bytecode, and `vm.store` to set an storage slots. Lastly, we reset
+///         the account the contract was initially deployed so it's not included in the `vm.dumpState`.
+contract L2Genesis is Script, Artifacts {
+    uint256 constant PROXY_COUNT = 2048;
+    uint256 constant PRECOMPILE_COUNT = 256;
+    DeployConfig public constant cfg =
+        DeployConfig(address(uint160(uint256(keccak256(abi.encode("optimism.deployconfig"))))));
+
+    /// @notice The storage slot that holds the address of a proxy implementation.
+    /// @dev `bytes32(uint256(keccak256('eip1967.proxy.implementation')) - 1)`
+    bytes32 internal constant PROXY_IMPLEMENTATION_ADDRESS =
+        0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;
+
+    /// @notice The storage slot that holds the address of the owner.
+    /// @dev `bytes32(uint256(keccak256('eip1967.proxy.admin')) - 1)`
+    bytes32 internal constant PROXY_ADMIN_ADDRESS = 0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103;
+
+    string internal outfile;
+
+    /// @dev Reads the deploy config, sets `outfile` which is where the `vm.dumpState` will be saved to, and
+    ///      loads in the addresses for the L1 contract deployments.
+    function setUp() public override {
+        Artifacts.setUp();
+
+        string memory path = string.concat(vm.projectRoot(), "/deploy-config/", deploymentContext, ".json");
+        vm.etch(address(cfg), vm.getDeployedCode("DeployConfig.s.sol:DeployConfig"));
+        vm.label(address(cfg), "DeployConfig");
+        vm.allowCheatcodes(address(cfg));
+        cfg.read(path);
+
+        outfile = string.concat(vm.projectRoot(), "/deployments/", deploymentContext, "/genesis-l2.json");
+
+        _loadAddresses(string.concat(vm.projectRoot(), "/deployments/", deploymentContext, "/.deploy"));
+    }
+
+    /// @dev Sets the precompiles, proxies, and the implementation accounts to be `vm.dumpState`
+    ///      to generate a L2 genesis alloc.
+    /// @notice The alloc object is sorted numerically by address.
+    function run() public {
+        _setPrecompiles();
+        _setProxies();
+        _setImplementations();
+
+        /// Reset so its not included state dump
+        vm.etch(address(cfg), "");
```

Comment:
I don't think this is sufficient, I still see `0x9568d36E291c2C4c34fa5593fcE73715abEf6F9c` (deploy config) in the genesis file because it has storage slots set. Some other things look off in the JSON output:
- `0x2e234dae75c793f67a35089c9d99245e1c58470b` (SafeSingleton) has code of `0x`. 
- `0x5615deb798bb3e4dfa0139dfa1b3d433cc23b72f` (SafeProxyFactory) has no state.

We should add an `L2Genesis.t.sol` with tests that verify the resulting genesis file has the properties we'd expect. Scripts are critical code and should always be tested. You could abstract the assertions so they can be run as standalone tests (`forge test --mt test/scripts/L2Genesis.t.sol`), and so they are called at the end of the script (when running `forge script scripts/L2Genesis.s.sol`) to verify the output file

One EVM invariant we can test for that would have caught this is: no account in the genesis JSON should have storage set without any other state 

Another test is that no account should be present if it has no state.

Then we can also test that accounts are numerically sorted, the expected addresses exist, etc

Reply:
@mds1 What version of `forge` do you have installed? Mark has [merged a required change](https://github.com/ethereum-optimism/optimism/pull/9313) in order to remove these "empty" accounts. I have the following versions installed and don't see the address included in the genesis output:

Sidenote, is there a way to enforce the version of `forge` installed for the monorepo to run the script? Seems like a really easy bug to have slip by

```
foundryup: installed - forge 0.2.0 (7922fd5 2024-02-05T00:19:38.726921000Z)
foundryup: installed - cast 0.2.0 (7922fd5 2024-02-05T00:19:38.726592000Z)
foundryup: installed - anvil 0.2.0 (7922fd5 2024-02-05T00:19:38.713486000Z)
foundryup: installed - chisel 0.2.0 (7922fd5 2024-02-05T00:19:38.704298000Z)
```

<img width="558" alt="image" src="https://github.com/ethereum-optimism/optimism/assets/20375610/3163c238-5c32-4f14-b71d-93476264d310">

<img width="554" alt="image" src="https://github.com/ethereum-optimism/optimism/assets/20375610/d59f5b16-3ea2-490f-ae99-debbb36eb5cd">

<img width="550" alt="image" src="https://github.com/ethereum-optimism/optimism/assets/20375610/242933bb-a387-44a8-a815-b2e88ca32ef1">

Reply:
Forge uses nightly releases, so they are all version `0.2.0`, and it's the commit hash that uniquely identifies them. I'm pretty sure forge does not support a `version` key in the config file to enforce a version.

If we need to enforce a forge version in a script, we could write a helper method in solidity that compares the hash logged in `forge --version` with the hash specified in our `versions.json`

Reply:
Ok, confirmed that by bumping to the updated forge version I no longer see those addresses. I do still think we should add tests, which will help prevent any regressions and ensure we didn't miss anything with manual review

Reply:
💯 Working on adding tests now

> If we need to enforce a forge version in a script, we could write a helper method in solidity that compares the hash logged in forge --version with the hash specified in our versions.json

I feel like we should move forward with this, it would be too easy for someone to not have updated Forge before running the L2 genesis script, and produce a malformed `allloc`

Reply:
There is already some existing test coverage in `Predeploys.t.sol`. We could port the existing Go tests to solidity as well

Reply:
There is no way to configure the version of forge via `foundry.toml`. We do have a script that verifies the foundry install as part of `contracts-bedrock` but it makes no assertion on version, that would be the place to add the check

Reply:
#9410 Adds test for this PR

Reply:
> We do have a script that verifies the foundry install as part of contracts-bedrock but it makes no assertion on version, that would be the place to add the check

#9429 addresses this